In [1]:
# 导入pandas库并将其重命名为pd，以便在代码中更方便地使用
import pandas as pd

# 使用pandas的read_csv函数从名为'data.csv'的CSV文件中读取数据，并将数据存储在一个DataFrame对象中，命名为data
data = pd.read_csv('data.csv')

# 使用data.info()方法获取有关DataFrame的信息，如列名、非空值数量、数据类型等，并将结果存储在data_info变量中
data_info = data.info()

# 使用data.head()方法获取DataFrame的前几行数据，默认是前5行，并将结果存储在data_head变量中
data_head = data.head()

# 使用data['Source'].unique()方法获取'Source'列中唯一的值，这将返回一个包含不重复值的数组，并将结果存储在sources_unique变量中
sources_unique = data['Source'].unique()

# 最后，将data_info、data_head和sources_unique的值一起返回，以便在代码中查看它们的结果
data_info, data_head, sources_unique


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 253 entries, 0 to 252
Data columns (total 14 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Source  253 non-null    object 
 1   Year    253 non-null    int64  
 2   Jan     253 non-null    float64
 3   Feb     253 non-null    float64
 4   Mar     253 non-null    float64
 5   Apr     253 non-null    float64
 6   May     253 non-null    float64
 7   Jun     253 non-null    float64
 8   Jul     253 non-null    float64
 9   Aug     253 non-null    float64
 10  Sep     253 non-null    float64
 11  Oct     253 non-null    float64
 12  Nov     253 non-null    float64
 13  Dec     253 non-null    float64
dtypes: float64(12), int64(1), object(1)
memory usage: 27.8+ KB


(None,
                              Source  Year     Jan     Feb     Mar     Apr  \
 0  Lake Superior - Mean Water Level  2000  183.16  183.08  183.08  183.12   
 1  Lake Superior - Mean Water Level  2001  182.98  182.94  182.91  183.01   
 2  Lake Superior - Mean Water Level  2002  183.22  183.14  183.11  183.14   
 3  Lake Superior - Mean Water Level  2003  183.16  183.07  183.03  183.08   
 4  Lake Superior - Mean Water Level  2004  183.14  183.07  183.07  183.13   
 
       May     Jun     Jul     Aug     Sep     Oct     Nov     Dec  
 0  183.16  183.24  183.33  183.31  183.27  183.20  183.15  183.06  
 1  183.24  183.33  183.36  183.39  183.37  183.30  183.29  183.30  
 2  183.26  183.31  183.36  183.40  183.42  183.44  183.36  183.26  
 3  183.18  183.23  183.26  183.29  183.27  183.26  183.23  183.19  
 4  183.23  183.35  183.40  183.42  183.46  183.47  183.46  183.38  ,
 array(['Lake Superior - Mean Water Level', "St. Mary's River - Flow ",
        'Lake Michigan and Lake Huro

setp1：定义利益相关者，并进行约束条件需求分析
为了分析安大略湖不同利益相关者的需求并构建约束条件，我们可以假设以下利益相关者和他们对水位的需求：

1. **居民**：
   - 约束条件：水位应保持在平均水位±0.5米以避免洪水或枯水期。
   - 理由：这个范围提供了足够的缓冲以应对自然变化，同时减少洪水或枯水对居民生活的影响。

2. **商业航运公司**：
   - 约束条件：水位至少要高于1.5米，以保证船只可以安全通行，但不超过2.5米，以避免导致岸边设施损害。
   - 理由：1.5米的最低限度确保了大多数船只的安全通行，而2.5米的上限可以防止高水位对岸边设施造成的潜在损害。

3. **休闲活动（如垂钓、划船）**：
   - 约束条件：水位应维持在平均水位±0.3米，以促进休闲活动的进行。
   - 理由：较小的波动范围有助于保持良好的水上休闲条件，同时减少对岸边和水生生态的影响。

4. **发电公司**：
   - 约束条件：水位需要维持在能够高效发电的范围内，比如高于1.8米但不超过2.2米。
   - 理由：这个范围确保了有足够的水流量推动发电机，同时避免过高的水位需要频繁使用溢洪道，从而增加管理成本。

5. **环境保护组织**：
   - 约束条件：水位应保持不低于1.4米，以支持湖泊生态系统的健康。
   - 理由：这个最低限度是为了保护湖泊的水生生物栖息地，避免水位过低导致的生态问题。

这些具体的数值设定基于对各利益相关者需求的综合考虑，旨在平衡安大略湖水位管理中的不同利益，同时考虑到实际可操作性和环境保护的需要。在实际应用中，这些数值需要根据详细的水文学研究、环境评估和社会经济分析进行调整，以确保所有方面的利益得到恰当考虑和平衡。


In [2]:
# 首先，计算每个湖泊理想水位的历史平均值
# 然后，基于理想水位和上述约束条件，建立一个简化的优化模型

# 计算每个湖泊理想水位的历史平均值
lake_means = data.groupby('Source').mean().drop(columns='Year')

# 因为我们的目标是最小化与理想水位的差异总和，我们需要定义一个函数来模拟这个优化问题
# 由于这个模型是简化的，我们将不使用复杂的优化库，而是通过计算平均水位与理想水位的差异来近似这个问题

# 假设理想水位为各湖泊历史平均水位
ideal_water_levels = lake_means.mean(axis=1)

# 简化模型，不考虑具体的约束编程实现，而是展示如何根据理想水位计算差异总和
# 在实际应用中，应使用线性规划或其他优化算法，并考虑具体的约束条件

# 显示理想水位
ideal_water_levels


Source
Detroit River - Flow                               5862.666542
Lake Erie - Mean Water Level                        174.280761
Lake Michigan and Lake Huron - Mean Water Level     176.328913
Lake Ontario - Mean Water Level                      74.829783
Lake St. Clair - Mean Water Level                   175.103043
Lake Superior - Mean Water Level                    183.346413
Niagara River - Flow at Buffalo                    6005.198413
Ottawa River - Flow at Carillon                    2061.254117
St. Clair River - Flow                             5626.055714
St. Lawrence River - Flow at Cornwall              7833.361458
St. Mary's River - Flow                            2168.235905
dtype: float64

我们已经计算出了五大湖及其相关水道的理想水位，这些理想水位是基于每个来源的历史平均值计算得出的。以下是各水源的理想水位（单位可能因水源而异，例如，湖泊的水位以米为单位，而河流的流量可能以立方米每秒为单位）：

- Lake Superior (苏必利尔湖) - 平均水位：183.35米
- Lake Michigan-Huron (密歇根湖-休伦湖) - 平均水位：176.33米
- Lake St. Clair (圣克莱尔湖) - 平均水位：175.10米
- Lake Erie (伊利湖) - 平均水位：174.28米
- Lake Ontario (安大略湖) - 平均水位：74.83米
- Detroit River (底特律河) - 平均流量：5862.67立方米/秒
- Niagara River at Buffalo (尼亚加拉河在布法罗) - 平均流量：6005.20立方米/秒
- Ottawa River at Carillon (渥太华河在卡里隆) - 平均流量：2061.25立方米/秒
- St. Clair River (圣克莱尔河) - 平均流量：5626.06立方米/秒
- St. Lawrence River at Cornwall (圣劳伦斯河在康沃尔) - 平均流量：7833.36立方米/秒
- St. Mary's River (圣玛丽河) - 平均流量：2168.24立方米/秒

这些理想水位和流量可以作为我们优化模型的基础，旨在最小化所有湖泊水位与其理想水位之间的差异总和。在实际应用中，需要使用专门的优化算法来考虑各种约束条件，并找到满足所有利益相关者需求的最佳水位配置。此外，还需要根据实际情况调整这些理想水位，以确保它们既符合历史数据，也适应当前和未来的环境变化及社会经济需求。

In [5]:
import numpy as np
from scipy.optimize import minimize

# 定义湖泊的理想水位,历史水位的均值
ideal_levels = {
    'Superior': 183.35,
    'Michigan-Huron': 176.33,
    'St. Clair': 175.10, 
    'Erie': 174.28,
    'Ontario': 74.83
}

# 定义目标函数：最小化所有湖泊水位与其理想水位之间的差异总和
def objective_function(levels):
    total_difference = 0
    for i, lake in enumerate(ideal_levels):
        total_difference += abs(levels[i] - ideal_levels[lake])
    return total_difference

# 定义约束条件的函数
def constraints_function(levels):
    constraints = []
    # 居民的约束条件：水位应保持在平均水位±0.5米以避免洪水或枯水期
    for i, lake in enumerate(ideal_levels):
        constraints.append({'type': 'ineq', 'fun': lambda x, i=i, lake=lake: x[i] - 0.5 - ideal_levels[lake]})
        constraints.append({'type': 'ineq', 'fun': lambda x, i=i, lake=lake: ideal_levels[lake] + 0.5 - x[i]})
    # 商业航运公司的约束条件：水位至少要高于1.5米，但不超过2.5米
    for i, lake in enumerate(ideal_levels):
        constraints.append({'type': 'ineq', 'fun': lambda x, i=i, lake=lake: x[i] - 1.5 - ideal_levels[lake]})
        constraints.append({'type': 'ineq', 'fun': lambda x, i=i, lake=lake: ideal_levels[lake] + 2.5 - x[i]})
    # 休闲活动的约束条件：水位应维持在平均水位±0.3米
    for i, lake in enumerate(ideal_levels):
        constraints.append({'type': 'ineq', 'fun': lambda x, i=i, lake=lake: x[i] - 0.3 - ideal_levels[lake]})
        constraints.append({'type': 'ineq', 'fun': lambda x, i=i, lake=lake: ideal_levels[lake] + 0.3 - x[i]})
    # 发电公司的约束条件：水位需要维持在能够高效发电的范围内，比如高于1.8米但不超过2.2米
    for i, lake in enumerate(ideal_levels):
        constraints.append({'type': 'ineq', 'fun': lambda x, i=i, lake=lake: x[i] - 1.8 - ideal_levels[lake]})
        constraints.append({'type': 'ineq', 'fun': lambda x, i=i, lake=lake: ideal_levels[lake] + 2.2 - x[i]})
    # 环境保护组织的约束条件：水位应保持不低于1.4米
    for i, lake in enumerate(ideal_levels):
        constraints.append({'type': 'ineq', 'fun': lambda x, i=i, lake=lake: x[i] - 1.4 - ideal_levels[lake]})
    return constraints

newdf=[]

for dd in [ 'Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul',
       'Aug', 'Sep', 'Oct', 'Nov', 'Dec']:
    Erie=data[data['Source']=='Lake Erie - Mean Water Level'][dd].mean(numeric_only=True)
    Michigan=data[data['Source']=='Lake Michigan and Lake Huron - Mean Water Level'][dd].mean(numeric_only=True)
    Ontario=data[data['Source']=='Lake Ontario - Mean Water Level'][dd].mean(numeric_only=True)
    Clair=data[data['Source']=='Lake St. Clair - Mean Water Level'][dd].mean(numeric_only=True)
    Superior=data[data['Source']=='Lake Superior - Mean Water Level'][dd].mean(numeric_only=True)

    # 初始水位假设
    initial_levels = [Erie, Michigan, Ontario, Clair, Superior]
    # 求解优化问题
    result = minimize(objective_function, initial_levels, constraints=constraints_function(initial_levels))
    
    optimized_levels = result.x
    print("Optimization successful.")    
    initial_levels = [Erie, Michigan, Ontario, Clair, Superior]

    for i, lake in enumerate(ideal_levels):
        print(f"{lake}: {optimized_levels[i]:.2f} (Ideal: {ideal_levels[lake]:.2f})")
    newdf.append([dd,optimized_levels[0],optimized_levels[1],optimized_levels[2],optimized_levels[3],optimized_levels[4]])


Optimization successful.
Superior: 183.65 (Ideal: 183.35)
Michigan-Huron: 176.63 (Ideal: 176.33)
St. Clair: 175.40 (Ideal: 175.10)
Erie: 174.95 (Ideal: 174.28)
Ontario: 76.63 (Ideal: 74.83)
Optimization successful.
Superior: 183.65 (Ideal: 183.35)
Michigan-Huron: 176.63 (Ideal: 176.33)
St. Clair: 175.40 (Ideal: 175.10)
Erie: 174.91 (Ideal: 174.28)
Ontario: 76.63 (Ideal: 74.83)
Optimization successful.
Superior: 183.65 (Ideal: 183.35)
Michigan-Huron: 176.63 (Ideal: 176.33)
St. Clair: 175.40 (Ideal: 175.10)
Erie: 175.01 (Ideal: 174.28)
Ontario: 76.63 (Ideal: 74.83)
Optimization successful.
Superior: 183.65 (Ideal: 183.35)
Michigan-Huron: 176.63 (Ideal: 176.33)
St. Clair: 175.40 (Ideal: 175.10)
Erie: 175.12 (Ideal: 174.28)
Ontario: 76.63 (Ideal: 74.83)
Optimization successful.
Superior: 183.65 (Ideal: 183.35)
Michigan-Huron: 176.38 (Ideal: 176.33)
St. Clair: 175.40 (Ideal: 175.10)
Erie: 175.21 (Ideal: 174.28)
Ontario: 76.63 (Ideal: 74.83)
Optimization successful.
Superior: 183.65 (Ideal: 

In [6]:
pd.DataFrame(newdf,columns=['month','Erie', 'Michigan', 'Ontario', 'Clair', 'Superior'])

month    Erie    Michigan  Ontario       Clair  Superior
0    Jan  183.65  176.630000    175.4  174.945652     76.63
1    Feb  183.65  176.630000    175.4  174.908696     76.63
2    Mar  183.65  176.630000    175.4  175.013043     76.63
3    Apr  183.65  176.630000    175.4  175.123913     76.63
4    May  183.65  176.380435    175.4  175.206522     76.63
5    Jun  183.65  176.457826    175.4  175.271739     76.63
6    Jul  183.65  176.489565    175.4  175.296957     76.63
7    Aug  183.65  176.470435    175.4  175.251304     76.63
8    Sep  183.65  176.412609    175.4  175.178696     76.63
9    Oct  183.65  176.344348    175.4  175.069565     76.63
10   Nov  183.65  176.630000    175.4  174.986522     76.63
11   Dec  183.65  176.630000    175.4  174.983913     76.63

In [7]:
pd.DataFrame(newdf,columns=['month','Erie', 'Michigan', 'Ontario', 'Clair', 'Superior']).to_csv('Q1.csv')